fix: clone object which have functions

airman5573 opened the pull request:
Above all, I don't think I understand the context of this library properly. **If there is any problem that my code will cause, please let me know**.

I think you decided not to just clone an object that contains a function in order not to create a new File object(maybe..?). But I think it's an error that when a user puts a function in defaultValues, the object itself containing the function is not copied.

So I changed cloneObject like below.

```javascript
let copy: any;
const isArray = Array.isArray(data);

if (data instanceof Date) {
  copy = new Date(data);
} else if (data instanceof Set) {
  copy = new Set(data);
} else if (globalThis.Blob && data instanceof Blob) { // check it is Blob
  copy = data;
} else if (isArray || isObject(data)) {
  copy = isArray ? [] : {};
  for (const key in data) { // dont check data[key] is function or not
    copy[key] = cloneObject(data[key]); // process clone
  }
} else {
  return data;
}

return copy;
```

This code prevent a bug(maybe..) below.

```javascript
function DefaultValues() {
  const defaultValues = {
    testFunction: () => { console.log('this is function'); },
    test: 'test',
    checkbox: ['1', '2'],
    test1: {
      firstName: 'firstName',
      lastName: ['lastName0', 'lastName1'],
      deep: {
        nest: 'nest',
      },
    },
  }
  const { register, handleSubmit, getValues } = useForm<{
    testFunction: () => void;
    test: string;
    test1: {
      firstName: string;
      lastName: string[];
      deep: {
        nest: string;
      };
    };
    checkbox: NestedValue<string[]>;
  }>({
    defaultValues,
  });
  const [show, setShow] = React.useState(true);

  const onSubmit = (data: any) => {
    defaultValues.test = 'another-test-string'; // this change the result of `getValues()`;
    console.log(getValues());
  }

  return (
    <>
      {show ? (
        <form onSubmit={handleSubmit(onSubmit)}>
          <input {...register('test')} />
          <input {...register('test1.firstName')} />
          <input {...register('test1.deep.nest')} />
          <input {...register('test1.deep.nest')} />
          <input {...register('test1.lastName.0')} />
          <input {...register('test1.lastName.1')} />
          <input type="checkbox" value={'1'} {...register('checkbox')} />
          <input type="checkbox" value={'2'} {...register('checkbox')} />
          <button type="submit">SUBMIT</button>
        </form>
      ) : null}
      <button type={'button'} id={'toggle'} onClick={() => setShow(!show)}>
        toggle
      </button>
    </>
  );
}

export default DefaultValues;
```

If we skip clone an object which has a function, the result of `getValues().test` will be `another-test-string`.

airman5573 committed:
fix: clone object which have functions

now cloneObject function clone object which have functions.
And skip clone when it is instance of Blob.

bluebill1049 commented:
thanks for the PR. Please avoid the clone function as well, this is related to complex objects (which contain functions) such as draftJS or Moment.

airman5573 commented:
> thanks for the PR. Please avoid the clone function as well, this is related to complex objects (which contain functions) such as draftJS or Moment.

Thanks for checking !

How to avoid clone function ? 

I think there are two way to avoid clone function.  

1. throw error
the structuredClone function [throw an error](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#things_that_dont_work_with_structured_clone) when clone an object which contain functions.

2. clone other property except function.
```
let copy: any;
const isArray = Array.isArray(data);

if (data instanceof Date) {
  copy = new Date(data);
} else if (data instanceof Set) {
  copy = new Set(data);
} else if (isArray || isObject(data)) {
  copy = isArray ? [] : {};
  for (const key in data) {
    if (!isFunction(data[key])) {
      copy[key] = cloneObject(data[key]);
    }
  }
} else {
  return data;
}
```

Am I thinking right ? check please!

bluebill1049 commented:
That's awesome feedback above, yes we want to move into `structuredClone` eventually, for current I think your proposal 2 works well and fixes the bug.

airman5573 committed:
do not clone function in object
function is not serializable and this means that we cannot get the state of the functions's environment when it is copied.

airman5573 committed:
copy entier FileList object without deleting function
1. we cannot create FileList without file input
2. fileList in setValue.test.tsx is just a javascript implementation of FileList
3. so removed item function that we can pass the cloneObject test

bluebill1049 committed:
skip filelist and blob for the current usage coverage

bluebill1049 commented:
Thanks for the update, I am leaving blob and file list coverage out for now as this is not the original issue.

bluebill1049 committed:
Update cloneObject.ts

bluebill1049 committed:
Update cloneObject.ts

bluebill1049 committed:
Update cloneObject.ts

bluebill1049 commented:
Sorry to mess around with the commit, I am going to merge with this PR and have a play with the blob and filelist, ideally we can skip those checks.

bluebill1049 commented:
thanks very much for the issue report and help on the PR! ❤️
